utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+) -> hl.MatrixTable:
+    """
+    Computes a sample/interval MT with each entry containing the call rate for that sample/interval.
+    This can be used as input for imputing exome sequencing platforms.
+
+    :param MatrixTable mt: Input MT
+    :param Table intervals_ht: Table containing the intervals. This table has to be keyed by locus. Default is the Broad exome calling intervals.
+    :param bool bi_allelic_only: If set, only bi-allelic sites are used for the computation
+    :param bool autosomes_only: If set, only autosomal intervals are used.
+    :return: Callrate MT
+    :rtype: MatrixTable
+    """
+    logger.info('Computing call rate MatrixTable')
+    if autosomes_only:
+        callrate_mt = filter_to_autosomes(mt)
+
+    if bi_allelic_only:
+        callrate_mt = callrate_mt.filter_rows(bi_allelic_expr(callrate_mt))
+
+    callrate_mt = callrate_mt.annotate_rows(interval=intervals_ht[callrate_mt.locus][intervals_ht.key])
+    callrate_mt = callrate_mt.filter_rows(hl.is_defined(callrate_mt.interval))
+    callrate_mt = callrate_mt.select_entries(GT_not_called=hl.or_missing(hl.is_missing(callrate_mt.GT), hl.struct()))
+    callrate_mt = callrate_mt.group_rows_by(callrate_mt.interval).aggregate(callrate=hl.agg.fraction(hl.is_missing(callrate_mt.GT_not_called)))
+    return callrate_mt
+
+
+def run_platform_pca(
+        callrate_mt: hl.MatrixTable,
+        binzarization_threshold: Optional[float] =0.25
+) -> Tuple[List[float], hl.Table, hl.Table]:
+    """
+    Runs a PCA on a sample/interval MT with each entry containing the call rate.
+    When `binzarization_threshold ` is set, the callrate is transformed to a 0/1 value based on the thredhold.
+    E.g. with the default threshold of 0.25, all entries with a callrate < 0.25 are considered as 0s, others as 1s.
+
+    :param MatrixTable callrate_mt: Input callrate MT
+    :param float binzarization_threshold: binzarization_threshold. None is no threshold desired
+    :return: eigenvalues, scores_ht, loadings_ht
+    :rtype: (list of float, Table Table)
+    """
+    logger.info("Running platform PCA")
+
+    if binzarization_threshold is not None:
+        callrate_mt = callrate_mt.annotate_entries(callrate=hl.int(callrate_mt.callrate > binzarization_threshold))
+    # Center until Hail's PCA does it for you
+    callrate_mt = callrate_mt.annotate_rows(mean_callrate=hl.agg.mean(callrate_mt.callrate))
+    callrate_mt = callrate_mt.annotate_entries(callrate=callrate_mt.callrate - callrate_mt.mean_callrate)
+    eigenvalues, scores, loadings = hl.pca(callrate_mt.callrate, compute_loadings=True)  # TODO:  Evaluate whether computing loadings is a good / worthy thing
+    logger.info('Platform PCA eigenvalues: {}'.format(eigenvalues))
+
+    return eigenvalues, scores, loadings
+
+
+def assign_platform_from_pcs(
+        platform_pca_scores_ht: hl.Table,
+        pc_scores_ann: str = 'scores',
+        hdbscan_min_cluster_size: int = 500,
+        hdbscan_min_samples: int = None
```

Comment:
Is this the best default setting for `hdbscan_min_samples`? The package will set this value to match `hdbscan_min_cluster_size` by default if no value is provided, and this parameter controls how conservatively the algorithm assigns individual points to a cluster. The documentation says: ``` `min_samples` provides a measure of how conservative you want your clustering to be. The larger the value of `min_samples` you provide, the more conservative the clustering – more points will be declared as noise, and clusters will be restricted to progressively more dense areas.```

Reply:
So `None` -> default hdbscan setting, which I think makes sense. I've tried playing with this param (which is why I've added it as an argument), and ultimately just used the default. According the doc, they really recommend using only `min_cluster_size` if possible. Note that maybe we  should not provide a default for `min_cluser_size` (and maybe use something like `min(500, 0.1*n_samples)`  ?

Reply:
@gtiao : thoughts? (likely not super important, so if not I'm happy with leaving `500`)

Reply:
I like the option of scaling to sample size with `min(500, 0.1*n_samples)`. I'm fine with not messing around with the different `min_samples`; keeping it the same seems reasonable as a default behavior if we scale `min_cluster_size`. Great suggestion!